### PR TITLE
Support connect attributes and set the script name to `program_name` attribute if it is not specified.

### PR DESCRIPTION
**Pull request by kubo**
This is same feature with Perl DBD-MySQL `mysql_conn_attrs` option.
http://search.cpan.org/~capttofu/DBD-mysql-4.033/lib/DBD/mysql.pm#mysql_conn_attrs

### CODE REVIEW

File: ext/mysql2/client.c [hunk 1]
```diff
@@ -352,7 +352,36 @@ static VALUE rb_mysql_get_ssl_cipher(VALUE self)
   return rb_str;
 }
 
-static VALUE rb_connect(VALUE self, VALUE user, VALUE pass, VALUE host, VALUE port, VALUE database, VALUE socket, VALUE flags) {
+#if MYSQL_VERSION_ID >= 50606
```
Comment (sodabrew) [inline ext/mysql2/client.c]: I would prefer a have-function test on `mysql_options4` rather than a version check.

File: lib/mysql2/client.rb [hunk 2]
```diff
@@ -83,8 +88,11 @@ def initialize(opts = {})
       port = port.to_i unless port.nil?
       database = database.to_s unless database.nil?
       socket = socket.to_s unless socket.nil?
+      conn_attrs = conn_attrs.each_with_object({}) do |(key, value), hash|
+        hash[key.to_sym] = value.to_s
```
Comment (sodabrew) [inline lib/mysql2/client.rb]: If this were `hash[key.to_s] = value.to_s` and avoided symbols, it looks like a lot of code in `opt_connect_addr_add_i` could be removed, and just StringValueToCStr used for the key?

**Comment by sodabrew** [comment]
This looks awesome, thank you for the contribution! Review comments added.

**Comment by kubo** [comment]
I changed the code as you commented.

**Comment by sodabrew** [comment]
Merged!

**Pull request closed without merging**